README
## Encrypt PHP session data using files

The encryption is built using mcrypt extension 
and the randomness is managed by openssl
The default encryption algorithm is Rijndael-256
and we use CBC+HMAC (Encrypt-then-mac)

# WARNING

Due to the issue reported in [#19](https://github.com/ezimuel/PHP-Secure-Session/issues/19)
the current SecureSession handler implementation can be affected by **race-condition**
for concurrent session requests coming from the same SESSION_ID (for instance, with Ajax calls).

**I DO NOT SUGGEST TO USE THIS CODE IN PRODUCTION** until the issue [#19](https://github.com/ezimuel/PHP-Secure-Session/issues/19) will be  fixed!

# How to use it

Include the SecureSession.php in your project and use
the PHP session as usual.

# Demo

In the demo folder you can see a simple PHP script (demo.php)
that stores some data in $_SESSION and display the encrypted
data of the session file (stored in the temporary directory
of the operating system, e.g. /tmp in GNU/Linux).

  
Enrico Zimuel (enrico@zimuel.it)
Copyright GNU General Public License


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 1.1.0 - 2018-02-01

### Added

- [#29](https://github.com/ezimuel/PHP-Secure-Session/issues/29) added the
  `PHPSecureSession\Exception\AuthenticationFailedException` in case of
  authentication error of `SecureHandler::decrypt()`.

### Changed

- [d39d3aa](https://github.com/ezimuel/PHP-Secure-Session/commit/d39d3aa381ebc9cd98b6de59b34ec7129d43a8fd)
  refactored the unit test removing the usage of `@runInSeparateProcess`.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.


test/autoload.php
<?php
if (! interface_exists(\PHPUnit_Framework_Test::class)) {
    class_alias(\PHPUnit\Framework\Test::class, \PHPUnit_Framework_Test::class);
}
if (! class_exists(\PHPUnit_Framework_AssertionFailedError::class)) {
    class_alias(\PHPUnit\Framework\AssertionFailedError::class, \PHPUnit_Framework_AssertionFailedError::class);
}
if (! class_exists(\PHPUnit_Framework_TestSuite::class)) {
    class_alias(\PHPUnit\Framework\TestSuite::class, \PHPUnit_Framework_TestSuite::class);
}
if (! class_exists(\PHPUnit\Framework\Error\Error::class)) {
    class_alias(\PHPUnit_Framework_Error::class, \PHPUnit\Framework\Error\Error::class);
}
if (! class_exists(\PHPUnit\Framework\Error\Notice::class)) {
    class_alias(\PHPUnit_Framework_Error_Notice::class, \PHPUnit\Framework\Error\Notice::class);
}


test/bootstrap.php
<?php
chdir(dirname(__DIR__));
require 'vendor/autoload.php';
session_start();


test/demo/index.php
<?php
/**
 * Demo script for PHP-Secure-Session
 *
 * @author    Enrico Zimuel (enrico@zimuel.it)
 * @copyright MIT License
 */
$start = microtime(true);

ini_set('session.save_handler', 'files');

$autoload = __DIR__ . '/../../vendor/autoload.php';
if (! file_exists($autoload)) {
  echo "You need to execute <strong>composer install</strong>!";
  exit;
}
require_once $autoload;

// change the default session folder in a temporary dir
session_save_path(sys_get_temp_dir());
session_start();

if (empty($_SESSION['time'])) {
  $_SESSION['time'] = time(); // set the time
}
session_write_close();

$filename = sys_get_temp_dir() . DIRECTORY_SEPARATOR . 'sess_' . session_id();

$time = microtime(true) - $start;

echo "<h1>PHP-Secure-Session Demo</h1>";
echo "<p>Session created at <strong>" . date("G:i:s ", $_SESSION['time']) . "</strong></p>";
echo "<p>Session file: <strong>" . $filename . "</strong></p>";
echo "<p>Content:<br><pre>" . session_encode() . "</pre></p>";
echo "<p>Encrypted content in Base64:<br><pre>" . base64_encode(file_get_contents($filename)). "</pre></p>";
echo "<p><strong>Note:</strong> If you reload the page you will see the encrypted data changing</p>";

printf("Execution time: %.6f", $time * 1000);


demo/demo.php
<?php
/**
 * Demo script for SecureSession
 * 
 * Requirement: PHP 5.2.1+
 *
 * @author    Enrico Zimuel (enrico@zimuel.it)
 * @copyright GNU General Public License 
 */
require_once '../SecureSession.php';

// change the default session folder in a temporary dir
$sessionPath = sys_get_temp_dir();
session_save_path($sessionPath);
session_start();
 if (empty($_SESSION['time'])) {
     $_SESSION['time'] = time(); // set the time
     $_SESSION['JIJO'] = 'HAI HOW ARE YOU';

   }
   $filename = $sessionPath . DIRECTORY_SEPARATOR . session_name() . '_' . session_id();

   echo "<h1>SecureSession Demo</h1>";
   echo "<br>Session created at <strong>" . date("G:i:s ", $_SESSION['time']) . "</strong>";
    echo "<br>Session file: <strong>" . $filename . "</strong>";
    echo "<br><br>Encrypted content:<br><pre>" . file_get_contents($filename). "</pre>";
    echo "<br><strong>Note:</strong> If you reload the page you see the encrypted data change each time";


src/SecureHandler.php
<?php
/**
 * Encrypt PHP session data for the internal PHP save handlers
 *
 * The encryption is built using OpenSSL extension with AES-256-CBC and the
 * authentication is provided using HMAC with SHA256.
 *
 * @author    Enrico Zimuel (enrico@zimuel.it)
 * @copyright MIT License
 */
namespace PHPSecureSession;

use SessionHandler;

class SecureHandler extends SessionHandler
{
    /**
     * Encryption and authentication key
     * @var string
     */
    protected $key;

    /**
     * Constructor
     */
    public function __construct()
    {
        if (! extension_loaded('openssl')) {
            throw new \RuntimeException(sprintf(
                "You need the OpenSSL extension to use %s",
                __CLASS__
            ));
        }
        if (! extension_loaded('mbstring')) {
            throw new \RuntimeException(sprintf(
                "You need the Multibytes extension to use %s",
                __CLASS__
            ));
        }
    }

    /**
     * Open the session
     *
     * @param string $save_path
     * @param string $session_name
     * @return bool
     */
    public function open($save_path, $session_name)
    {
        $this->key = $this->getKey('KEY_' . $session_name);
        return parent::open($save_path, $session_name);
    }

    /**
     * Read from session and decrypt
     *
     * @param string $id
     */
    public function read($id)
    {
        $data = parent::read($id);
        return empty($data) ? '' : $this->decrypt($data, $this->key);
    }

    /**
     * Encrypt the data and write into the session
     *
     * @param string $id
     * @param string $data
     */
    public function write($id, $data)
    {
        return parent::write($id, $this->encrypt($data, $this->key));
    }

    /**
     * Encrypt and authenticate
     *
     * @param string $data
     * @param string $key
     * @return string
     */
    protected function encrypt($data, $key)
    {
        $iv = random_bytes(16); // AES block size in CBC mode
        // Encryption
        $ciphertext = openssl_encrypt(
            $data,
            'AES-256-CBC',
            mb_substr($key, 0, 32, '8bit'),
            OPENSSL_RAW_DATA,
            $iv
        );
        // Authentication
        $hmac = hash_hmac(
            'SHA256',
            $iv . $ciphertext,
            mb_substr($key, 32, null, '8bit'),
            true
        );
        return $hmac . $iv . $ciphertext;
    }

    /**
     * Authenticate and decrypt
     *
     * @param string $data
     * @param string $key
     * @return string
     */
    protected function decrypt($data, $key)
    {
        $hmac       = mb_substr($data, 0, 32, '8bit');
        $iv         = mb_substr($data, 32, 16, '8bit');
        $ciphertext = mb_substr($data, 48, null, '8bit');
        // Authentication
        $hmacNew = hash_hmac(
            'SHA256',
            $iv . $ciphertext,
            mb_substr($key, 32, null, '8bit'),
            true
        );
        if (! hash_equals($hmac, $hmacNew)) {
            throw new Exception\AuthenticationFailedException('Authentication failed');
        }
        // Decrypt
        return openssl_decrypt(
            $ciphertext,
            'AES-256-CBC',
            mb_substr($key, 0, 32, '8bit'),
            OPENSSL_RAW_DATA,
            $iv
        );
    }

    /**
     * Get the encryption and authentication keys from cookie
     *
     * @param string $name
     * @return string
     */
    protected function getKey($name)
    {
        if (empty($_COOKIE[$name])) {
            $key         = random_bytes(64); // 32 for encryption and 32 for authentication
            $cookieParam = session_get_cookie_params();
            $encKey      = base64_encode($key);
            setcookie(
                $name,
                $encKey,
                // if session cookie lifetime > 0 then add to current time
                // otherwise leave it as zero, honoring zero's special meaning
                // expire at browser close.
                ($cookieParam['lifetime'] > 0) ? time() + $cookieParam['lifetime'] : 0,
                $cookieParam['path'],
                $cookieParam['domain'],
                $cookieParam['secure'],
                $cookieParam['httponly']
            );
            $_COOKIE[$name] = $encKey;
        } else {
            $key = base64_decode($_COOKIE[$name]);
        }
        return $key;
    }
}


SecureSession.php
<?php
/**
 * ------------------------------------------------
 * Encrypt PHP session data using files
 * ------------------------------------------------
 * The encryption is built using mcrypt extension
 * and the randomness is managed by openssl
 * The default encryption algorithm is AES (Rijndael-128)
 * and we use CBC+HMAC (Encrypt-then-mac) with SHA-256
 *
 * @author    Enrico Zimuel (enrico@zimuel.it)
 * @copyright GNU General Public License
 */
class SecureSession {
    /**
     * Encryption algorithm
     *
     * @var string
     */
    protected $_algo= MCRYPT_RIJNDAEL_128;
    /**
     * Key for encryption/decryption
    *
    * @var string
    */
    protected $_key;
    /**
     * Key for HMAC authentication
    *
    * @var string
    */
    protected $_auth;
    /**
     * Path of the session file
     *
     * @var string
     */
    protected $_path;
    /**
     * Session name (optional)
     *
     * @var string
     */
    protected $_name;
    /**
     * Size of the IV vector for encryption
     *
     * @var integer
     */
    protected $_ivSize;
    /**
     * Cookie variable name of the encryption + auth key
     *
     * @var string
     */
    protected $_keyName;
    /**
     * Generate a random key using openssl
     * fallback to mcrypt_create_iv
     *
     * @param  integer $length
     * @return string
     */
    protected function _randomKey($length=32) {
      if(function_exists('openssl_random_pseudo_bytes')) {
        $rnd = openssl_random_pseudo_bytes($length, $strong);
        if ($strong === true) {
          return $rnd;
        }
      }
      return mcrypt_create_iv($length, MCRYPT_DEV_URANDOM);
    }
    /**
     * Constructor
     */
    public function __construct()
    {
      session_set_save_handler(
        array($this, "open"),
        array($this, "close"),
        array($this, "read"),
        array($this, "write"),
        array($this, "destroy"),
        array($this, "gc")
      );
      if (!extension_loaded('mcrypt')) {
        throw new Exception("The SecureSession class needs the Mcrypt PHP extension, please install it.");
      }
    }
    /**
     * Open the session
     *
     * @param  string $save_path
     * @param  string $session_name
     * @return bool
     */
    public function open($save_path, $session_name)
    {
      $this->_path    = $save_path.'/';
	    $this->_name    = $session_name;
	    $this->_keyName = "KEY_$session_name";
	    $this->_ivSize  = mcrypt_get_iv_size($this->_algo, MCRYPT_MODE_CBC);

	    if (empty($_COOKIE[$this->_keyName]) || strpos($_COOKIE[$this->_keyName],':')===false) {
        $keyLength    = mcrypt_get_key_size($this->_algo, MCRYPT_MODE_CBC);
        $this->_key   = self::_randomKey($keyLength);
        $this->_auth  = self::_randomKey(32);
        $cookie_param = session_get_cookie_params();
        setcookie(
          $this->_keyName,
          base64_encode($this->_key) . ':' . base64_encode($this->_auth),
          // if session cookie lifetime > 0 then add to current time; otherwise leave it as zero, honoring zero's special meaning: expire at browser close.
          ($cookie_param['lifetime'] > 0) ? time() + $cookie_param['lifetime'] : 0,
          $cookie_param['path'],
          $cookie_param['domain'],
          $cookie_param['secure'],
          $cookie_param['httponly']
        );
	    } else {
        list ($this->_key, $this->_auth) = explode (':',$_COOKIE[$this->_keyName]);
        $this->_key  = base64_decode($this->_key);
        $this->_auth = base64_decode($this->_auth);
      }
	    return true;
    }
    /**
     * Close the session
     *
     * @return bool
     */
    public function close()
    {
      return true;
    }
    /**
     * Read and decrypt the session
     *
     * @param  integer $id
     * @return string
     */
    public function read($id)
    {
      $sess_file = $this->_path.$this->_name."_$id";
      if (!file_exists($sess_file)) {
        return false;
      }
  	  $data      = file_get_contents($sess_file);
      list($hmac, $iv, $encrypted)= explode(':',$data);
      $iv        = base64_decode($iv);
      $encrypted = base64_decode($encrypted);
      $newHmac   = hash_hmac('sha256', $iv . $this->_algo . $encrypted, $this->_auth);
      if ($hmac !== $newHmac) {
        return false;
      }
  	  $decrypt = mcrypt_decrypt(
        $this->_algo,
        $this->_key,
        $encrypted,
        MCRYPT_MODE_CBC,
        $iv
      );
      return rtrim($decrypt, "\0");
    }
    /**
     * Encrypt and write the session
     *
     * @param integer $id
     * @param string $data
     * @return bool
     */
    public function write($id, $data)
    {
      $sess_file = $this->_path . $this->_name . "_$id";
	    $iv        = mcrypt_create_iv($this->_ivSize, MCRYPT_DEV_URANDOM);
      $encrypted = mcrypt_encrypt(
        $this->_algo,
        $this->_key,
        $data,
        MCRYPT_MODE_CBC,
        $iv
      );
      $hmac  = hash_hmac('sha256', $iv . $this->_algo . $encrypted, $this->_auth);
      $bytes = file_put_contents($sess_file, $hmac . ':' . base64_encode($iv) . ':' . base64_encode($encrypted));
      return ($bytes !== false);
    }
    /**
     * Destoroy the session
     *
     * @param int $id
     * @return bool
     */
    public function destroy($id)
    {
      $sess_file = $this->_path . $this->_name . "_$id";
      setcookie ($this->_keyName, '', time() - 3600);
	    return(@unlink($sess_file));
    }
    /**
     * Garbage Collector
     *
     * @param int $max
     * @return bool
     */
    public function gc($max)
    {
    	foreach (glob($this->_path . $this->_name . '_*') as $filename) {
        if (filemtime($filename) + $max < time()) {
          @unlink($filename);
        }
  	  }
      return true;
    }
}

new SecureSession();
